### ISSUE TITLE
training loss too large

### ISSUE THREAD

**Issue by leonzgtee** [zju3dv/SMAP#13]
when i start training with this project, i find that the loss is too large to converge.  after 70k iteration, loss_2d loss about 384,loss_bone about 200,loss_root about 2.is there anything wrong with these loss. thanks.

**Comment by raypine** [comment]
After 40k iteration, loss_2d is about 110. loss_bone is about 1. loss_root is about 0.1.
The unit is cm in our setting. Is it correct ?

**Comment by minhpvo** [comment]
@leonzgtee  Is it possible for you to share the json conversion script from MuCo to SMAP format? Thanks!

**Comment by leonzgtee** [comment]
@raypine the unit previous uesed was mm, after changed to cm, it seems very close to the value. now the loss_2d: 157.297, loss_bone: 1.538, loss_root: 0.078 after 100k iteration's training.

**Comment by leonzgtee** [comment]
@minhpvo i modified it from lib/preprocess/create_annot.py, just as follows:
import numpy as np
import json
import sys
import os
import time
from collections import defaultdict

save_dir = '/SMAP-master/data/MuCo/annotations/'
ann_path = '/SMAP-master/lib/preprocess/MuCo-3DHP.json'
converted_ann_path = os.path.join(save_dir,'MuCo.json')
tic =time.time()
print('start loading MuCo json...')
dataset = json.load(open(ann_path,'r'))
print('Done t={}s'.format(time.time()-tic))
images = dict()
imgToAnns = defaultdict(list)
if 'images' in dataset:
    for img_info in dataset['images']:
        images[img_info['id']] = img_info
images_id = images.keys()
if 'annotations' in dataset:
    for ann in dataset['annotations']:
        imgToAnns[ann['image_id']].append(ann)
MuCo = [1,0,14,5,6,7,11,12,13,2,3,4,8,9,10]

def main():
    output_json = dict()
    output_json['root'] = []
    count = 0
    min_width = 1000
    min_height = 1000
    for img_id in images_id:
        bodys = list()
        # 加载images 部分内容
        img_info =images[img_id]
        anns = imgToAnns[img_id]
        for ann in anns:
            # if ann['num_keypoints'] < 3:
            #     continue
            body_2d = np.asarray(ann['keypoints_img'])
            body_3d = np.asarray(ann['keypoints_cam'])
            # body_vis = np.asarray(ann['keypoints_vis'])+0
            vis_num = np.array(ann['keypoints_vis']) + 0
            vis_num2 = np.logical_and(vis_num, np.array(21)) + 0
            body_vis = vis_num + vis_num2
            print(body_vis)

            body_new = np.zeros((15,11))
            for k in range(len(MuCo)):
                # if MuCo[k] < 0:
                #     continue
                body_new[k][0] = body_2d[MuCo[k]][0]
                body_new[k][1] = body_2d[MuCo[k]][1]
                body_new[k][2] = body_3d[MuCo[k]][2]/10
                body_new[k][3] = body_vis[MuCo[k]]
                body_new[k][4] = body_3d[MuCo[k]][0]/10
                body_new[k][5] = body_3d[MuCo[k]][1]/10
                body_new[k][6] = body_3d[MuCo[k]][2]/10
            body_new[:,7] = img_info['f'][0]
            body_new[:,8] = img_info['f'][1]
            body_new[:,9] = img_info['c'][0]
            body_new[:,10] = img_info['c'][1]
            bodys.append(body_new.tolist())
        if len(bodys) < 1:
            continue
        this_pic = dict()
        this_pic['dataset'] = 'MuCo'
        this_pic['img_paths'] = img_info['file_name']
        this_pic['img_width'] = img_info['width']
        this_pic['img_height'] = img_info['height']
        this_pic['image_id'] = img_info['id']
        this_pic['cam_id'] = 0
        this_pic['bodys'] = bodys
        this_pic['isValidation'] = 0
        output_json['root'].append(this_pic)
        count +=1
        min_width = min(min_width, img_info['width'])
        min_height = min(min_height, img_info['height'])
        print('writed {} with width {} and height {}'.format(img_info['file_name'], min_width, min_height))

    json.dump(output_json,open(converted_ann_path,'w'))
    print('Done! {} images converted'.format(count))
#

if __name__ == '__main__':
    main()

**Comment by leonzgtee** [comment]
@minhpvo i modified it from lib/preprocess/create_annot.py, just as follows:
import numpy as np
import json
import sys
import os
import time
from collections import defaultdict

save_dir = '/home/zhuguotao/oldsys/PycharmProjects/models_git/SMAP-master/data/MuCo/annotations/'
ann_path = '/home/zhuguotao/oldsys/PycharmProjects/models_git/SMAP-master/lib/preprocess/MuCo-3DHP.json'
converted_ann_path = os.path.join(save_dir,'MuCo11111111111111111111.json')
tic =time.time()
print('start loading MuCo json...')
dataset = json.load(open(ann_path,'r'))
print('Done t={}s'.format(time.time()-tic))
ann_type = ''
images = dict()
imgToAnns = defaultdict(list)
if 'images' in dataset:
    for img_info in dataset['images']:
        images[img_info['id']] = img_info
images_id = images.keys()
if 'annotations' in dataset:
    for ann in dataset['annotations']:
        imgToAnns[ann['image_id']].append(ann)
MuCo = [1,0,14,5,6,7,11,12,13,2,3,4,8,9,10]

def main():
    output_json = dict()
    output_json['root'] = []
    count = 0
    min_width = 1000
    min_height = 1000
    for img_id in images_id:
        bodys = list()
        # 加载images 部分内容
        img_info =images[img_id]
        anns = imgToAnns[img_id]
        for ann in anns:
            # if ann['num_keypoints'] < 3:
            #     continue
            body_2d = np.asarray(ann['keypoints_img'])
            body_3d = np.asarray(ann['keypoints_cam'])
            # body_vis = np.asarray(ann['keypoints_vis'])+0
            vis_num = np.array(ann['keypoints_vis']) + 0
            vis_num2 = np.logical_and(vis_num, np.array(21)) + 0
            body_vis = vis_num + vis_num2
            print(body_vis)

            body_new = np.zeros((15,11))
            for k in range(len(MuCo)):
                # if MuCo[k] < 0:
                #     continue
                print(k)
                body_new[k][0] = body_2d[MuCo[k]][0]
                body_new[k][1] = body_2d[MuCo[k]][1]
                body_new[k][2] = body_3d[MuCo[k]][2]/10
                body_new[k][3] = body_vis[MuCo[k]]
                body_new[k][4] = body_3d[MuCo[k]][0]/10
                body_new[k][5] = body_3d[MuCo[k]][1]/10
                body_new[k][6] = body_3d[MuCo[k]][2]/10
            # middle_shoulder_2d = (body_2d[2] + body_2d[5]) / 2
            # middle_shoulder_3d = (body_3d[2] +body_3d[5]) / 2
            # neck
            # body_new[0][0] = (middle_shoulder_2d[0] - body_2d[14][0]) * 0.185 + middle_shoulder_2d[0]
            # body_new[0][1] = (middle_shoulder_2d[1] - body_2d[14][1]) * 0.185 + middle_shoulder_2d[1]
            # body_new[0][3] = min(body_vis[14], body_vis[2], body_vis[5])
            # body_new[0][5] = middle_shoulder_3d[0]
            # body_new[0][6] = middle_shoulder_3d[1]
            # body_new[0][7] = middle_shoulder_3d[2]
            body_new[:,7] = img_info['f'][0]
            body_new[:,8] = img_info['f'][1]
            body_new[:,9] = img_info['c'][0]
            body_new[:,10] = img_info['c'][1]
            bodys.append(body_new.tolist())
        if len(bodys) < 1:
            continue
        this_pic = dict()
        this_pic['dataset'] = 'MuCo'
        this_pic['img_paths'] = img_info['file_name']
        this_pic['img_width'] = img_info['width']
        this_pic['img_height'] = img_info['height']
        this_pic['image_id'] = img_info['id']
        this_pic['cam_id'] = 0
        this_pic['bodys'] = bodys
        this_pic['isValidation'] = 0
        output_json['root'].append(this_pic)
        count +=1
        min_width = min(min_width, img_info['width'])
        min_height = min(min_height, img_info['height'])
        print('writed {} with width {} and height {}'.format(img_info['file_name'], min_width, min_height))

    json.dump(output_json,open(converted_ann_path,'w'))
    print('Done! {} images converted'.format(count))
#

if __name__ == '__main__':
    main()

**Comment by leonzgtee** [comment]
@minhpvo i modified it from lib/preprocess/create_annot.py, just as follows:
[create_MuCo_annot.txt](https://github.com/zju3dv/SMAP/files/5358360/create_MuCo_annot.txt)
